lua/bufferline/timing.lua

```diff
@@ -0,0 +1,90 @@
+local nvim = require'bufferline.nvim'
+local clock = require'bufferline.userclock'
+local active_buffer = -1
+local LAST_DECAY_TIME = 'bufferline_last_decay_time'
+local TIME_SCORE = 'bufferline_time_score'
+local TIME_ACTIVATED = 'bufferline_time_activated'
+
+local function is_listed(bufnr)
+  return nvim.buf_is_valid(bufnr) and nvim.buf_get_option(bufnr, 'buflisted')
+end
+
+local function get_last_decay_time(bufnr, now)
+  local ok, val = pcall(nvim.buf_get_var, bufnr, LAST_DECAY_TIME)
+  return ok and val or now
+end
+
+local function get_time_activated(bufnr, now)
+  local ok, val = pcall(nvim.buf_get_var, bufnr, TIME_ACTIVATED)
+  return ok and val or now
+end
+
+local function get_stored_score(bufnr)
+  local ok, val = pcall(nvim.buf_get_var, bufnr, TIME_SCORE)
+  if ok then
+    -- Floats get serialized as a table. See :help lua-special-tbl
+    if type(val) == 'table' then
+      return val[vim.val_idx]
+    else
+      return val
+    end
+  else
+    return 0
+  end
+end
+
+local function apply_decay(bufnr, now)
+  local last_time = get_last_decay_time(bufnr, now)
+  local delta = now - last_time
+  local score = get_stored_score(bufnr)
+
+  score = score / 2 ^ (delta / vim.g.bufferline.time_decay_rate)
+  nvim.buf_set_var(bufnr, TIME_SCORE, score)
+  nvim.buf_set_var(bufnr, LAST_DECAY_TIME, now)
+  return score
+end
+
+local function get_score(bufnr)
+  local now = clock.time()
+  if bufnr == active_buffer then
+    local time_activated = get_time_activated(bufnr, now)
+    return get_stored_score(bufnr) + (now - time_activated)
+  else
+    return apply_decay(bufnr, now)
+  end
+end
+
+local function on_leave_buffer()
+  if is_listed(active_buffer) then
+    local now = clock.time()
+    local score = get_stored_score(active_buffer)
+    local time_activated = get_time_activated(active_buffer, now)
+    local delta = now - time_activated
+    nvim.buf_set_var(active_buffer, TIME_SCORE, score + delta)
+    nvim.buf_set_var(active_buffer, LAST_DECAY_TIME, now)
+  end
+  active_buffer = -1
+end
+
+local function set_active_buffer(bufnr)
+  if bufnr ~= active_buffer then
+    on_leave_buffer()
+  end
+
+  if not is_listed(bufnr) or bufnr == active_buffer then
```

Comment:
Here you're using `is_listed` to determine if the buffer is present in the tabline, but I added an `exclude_ft` and `exclude_name` options to barbar recently, you'll need to create an `is_displayed` function to check for presence in the tabline.

See https://github.com/romgrk/barbar.nvim/blob/f677f1d93be06c41eaabc87d845cdd5bd46a259f/lua/bufferline/state.lua#L245-L267